### PR TITLE
Update proposal according to WG directions

### PR DESCRIPTION
**Pull request by alvestrand**
This prepares the document for adoption as a WG document.

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -353,37 +368,43 @@ to the track satisfies the track's constraints. Each track may receive a
...
-the {{MediaStreamTrackGenerator/writable}} of a {{MediaStreamTrackGenerator}}
-ends, its {{MediaStreamTrackGenerator/writable}} is [=WritableStream/closing|closed=].
 
 #### Constrainable properties #### {#generator-constrainable-properties}
 
-The following constrainable properties are defined for video
-{{MediaStreamTrackGenerator}}s and any {{MediaStreamTrack}}s sourced from
-a {{MediaStreamTrackGenerator}}:
+The following constrainable properties are defined for any {{MediaStreamTrack}}s sourced from
+a {{VideoTrackGenerator}}:
```
Comment (guidou) [inline index.bs:407]: VideoTrackGenerator -> VideoTrackSource

Comment (guidou) [reply index.bs:407]: Applies to other places below

Comment (alvestrand) [reply index.bs:407]: Thanks - at some point I unconsciously thought the name was VideoTrackGenerator. I actually like that name better than VideoTrackSource, but let's go with Jan-Ivar's name in this PR.

File: index.bs [hunk 2]
```diff
@@ -190,6 +190,10 @@ dictionary MediaStreamTrackProcessorInit {
 };
 </pre>
 
+Note: There is WG consensus that the interface should be exposed on DedicatedWorker.
+There is no WG consensus that the interface should not be exposed on Window.
```
Comment (jan-ivar) [inline index.bs:194]: I'm a bit concerned this double-negative is hard to parse. I would prefer something like. We normally don't require consensus to *not* do something, so it would seem more correct o say "There is no WG consensus that the interface should be exposed on Window.".

Or doing the same as for audio would be: "There is no WG consensus on whether or not the interface should be exposed on Window."

File: index.bs [hunk 3]
```diff
@@ -630,7 +589,199 @@ self.onmessage = async function(e) {
...
+<pre>
+
+More Javascript goes here
+
+</pre>
+
+# Considerations for sourcing audio tracks # {#audio-processor}
+
+These are kept in the document to allow further discussion of audio processing, if
+feasible. They do not indicate that the WG has consensus to add audio.
```
Comment (jan-ivar) [inline index.bs:721]: This section appears to normatively define/alter audio constraints in detail, which I'd consider surface/code. It feels out of place in a document that does not define an API for audio. The fact that this can be done without WebIDL seems like an artifact of our Constrainable pattern and our practice of expressing normative outcomes in tables instead of through algorithms.

Based on our agreement that we _"We document the parts that have consensus in code"_, I would expect this to be removed, unless it can be rewritten to be agnostic to kind.  cc @youennf

File: index.bs [hunk 4]
```diff
@@ -630,7 +589,199 @@ self.onmessage = async function(e) {
...
+      </td>
+      <td>
+        As a setting, this is the linear sample size of the latest {{AudioData}}
+        delivered by the track.
+      </td>
+    </tr>
+  </tbody>
+</table>
+
+## Example using audio/video co-processing
```
Comment (jan-ivar) [inline index.bs:784]: I think the two code examples dedicated to audio need to stay in alvestrand until we adopt an audio API.

Also, all examples are still main-thread based, and show the old API, which I think needs to be updated.

I provided 3 examples (3.1, 3.2. 3.3) in https://jan-ivar.github.io/mediacapture-transform/#examples I recommend we use to show how to transfer MST to a worker and using the API there.

**Review by jan-ivar — CHANGES_REQUESTED** [review]
Thanks for doing this merge!

I have some concerns around some WebIDL and some audio API surface that remains.

Also, the examples should be updated to show the worker model with transfer of MST. See link to my PR below which has these. But I can also do a separate PR for that if you prefer. LMK.

File: index.bs [hunk 5]
```diff
@@ -286,65 +290,76 @@ When the `[[track]]` of a {{MediaStreamTrackProcessor}} |processor|
...
-{{MediaStreamTrack/clone}} method. The {{WritableStream}} source is
-exposed as the {{MediaStreamTrackGenerator/writable}} field of
-{{MediaStreamTrackGenerator}}.
-
-Similarly to {{MediaStreamTrackProcessor}}, the {{WritableStream}} of
-an audio {{MediaStreamTrackGenerator}} accepts {{AudioData}} objects,
-and a video {{MediaStreamTrackGenerator}} accepts {{VideoFrame}} objects.
-When a {{VideoFrame}} or {{AudioData}} object is written to
-{{MediaStreamTrackGenerator/writable}},
+MediaStream model</a> in the dedicated worker environment. It has a two readonly
```
Comment (youennf) [inline index.bs:298]:
```suggestion
MediaStream model</a> in the dedicated worker environment. It has two readonly
```

File: index.bs [hunk 6]
```diff
@@ -286,65 +290,76 @@ When the `[[track]]` of a {{MediaStreamTrackProcessor}} |processor|
...
 
-dictionary MediaStreamTrackGeneratorInit {
-  required DOMString kind;
+
+### Interface definition ### {#video-source-interface}
+<pre class="idl">
+[Exposed=DedicatedWorker]
+interface VideoTrackSource {
+  constructor();
+  readonly attribute WritableStream writable;
```
Comment (youennf) [inline index.bs:322]: We have Promise<XYZ>, could we write it as WritableStream<VideoFrame>?

File: index.bs [hunk 7]
```diff
@@ -286,65 +290,76 @@ When the `[[track]]` of a {{MediaStreamTrackProcessor}} |processor|
...
+  VideoTrackSource()
 </dfn>
-1. If |init|.{{MediaStreamTrackGeneratorInit/kind}} is not `"audio"` or `"video"`,
-    throw a {{TypeError}}.
-2. Let |g| be a new {{MediaStreamTrackGenerator}} object.
-3. Initialize the {{MediaStreamTrack/kind}} field of |g| (inherited from {{MediaStreamTrack}})
-    with |init|.{{MediaStreamTrackGeneratorInit/kind}}.
-4. Return |g|.
+1. Let |source| be a new {{VideoTrackSource}} object.
+1. Let |track| be a new {{MediaStreamTrack}} object, whose {{MediaStreamTrack/kind}} is "video", and whose {{MediaStreamTrack/id}} is a new unique id generated by the user agent.
```
Comment (youennf) [inline index.bs:346]: We could be using https://w3c.github.io/mediacapture-main/#dfn-initialize-the-underlying-source here to tie source and MediaStreamTrack.

File: index.bs [hunk 8]
```diff
@@ -286,65 +290,76 @@ When the `[[track]]` of a {{MediaStreamTrackProcessor}} |processor|
...
+<dt><dfn attribute for=VideoTrackSource>writable</dfn></dt>
+<dd>Allows writing video frames to the {{VideoTrackSource}}. When this attribute
+is accessed for the first time, it MUST be initialized with the following steps:
+1. Initialize [=this=].{{VideoTrackSource/writable}} to be a new {{WritableStream}}.
+2. <a dfn for="WritableStream">Set up</a> [=this=].{{VideoTrackSource/writable}}, with its [=WritableStream/set up/writeAlgorithm=] set to [=writeFrame=] with |this| as parameter, with [=WritableStream/set up/closeAlgorithm=] set to [=closeWritable=] with |this| as parameter and [=WritableStream/set up/abortAlgorithm=] set to [=closeWritable=] with |this| as parameter.
+
+The <dfn>writeFrame</dfn> algorithm is given a |source| and a |frame| as input. It is defined by running the following steps:
+1. If |frame| is not a {{VideoFrame}} object, return [=a promise rejected with= a {{TypeError}}.
+1. If |source|.`[[isMuted]]` is false, send the media data backing |frame| to all live tracks sourced from |source|.
+1. Invoke the `close` method of |frame|.
```
Comment (youennf) [inline index.bs:361]: Seems good enough for now, I believe there might be cases where the close method will be called too late though so that JS could call close() before running that step for instance. Also, we should handle the case of closed frames.

File: index.bs [hunk 9]
```diff
@@ -353,37 +368,43 @@ to the track satisfies the track's constraints. Each track may receive a
...
-#### clone #### {#generator-clone}
-The {{MediaStreamTrack/clone}} method on a {{MediaStreamTrackGenerator}}
-returns a new {{MediaStreamTrack}} object whose source is the
-same as the one for the {{MediaStreamTrackGenerator}} being cloned.
-This source is the {{MediaStreamTrackGenerator/writable}} field of
-the {{MediaStreamTrackGenerator}}.
+#### stop #### {#video-source-stop}
+The {{MediaStreamTrack/stop}} method stops the track. When the last track
+sourced from a {{VideoTrackSource}} ends, that {{VideoTrackSource}}'s
+{{VideoTrackSource/writable}} is [=WritableStream/closing|closed=].
```
Comment (youennf) [inline index.bs:401]: We might want to call one of the writable stream spec algorithm

File: index.bs [hunk 10]
```diff
@@ -630,7 +589,199 @@ self.onmessage = async function(e) {
...
+to both a &lt;video&gt& tag and an {{RTCPeerConnection}}, attaching the resulting {{MediaStreamTrack}}
+to multiple sinks may be the most appropriate mechanism.
+
+For cases where the downstream processing takes frames, not streams, the frames can
+be cloned as needed and sent off to the downstream processing; "clone" is a cheap operation.
+
+When the stream is the output of some processing, and both branches need a Stream object
+to do further processing, one needs a function that produces two streams from one stream.
+
+However, the standard tee() operation is problematic
```
Comment (youennf) [inline index.bs:620]: Would be worth pointing to the streams issue.

File: index.bs [hunk 11]
```diff
@@ -630,7 +589,199 @@ self.onmessage = async function(e) {
...
+[Exposed=Window,DedicatedWorker]
+interface MediaStreamTrackGenerator : MediaStreamTrack {
+    constructor(MediaStreamTrackGeneratorInit init);
+    attribute WritableStream writable;  // VideoFrame or AudioFrame
+};
+
+dictionary MediaStreamTrackGeneratorInit {
+  required DOMString kind;
+};
+</pre>
```
[the remark this answers is not in the document]
Comment (youennf) [reply index.bs:698]: I agree that we should not try to document non-standard APIs here.

File: index.bs (hunk 3 above)
Comment (youennf) [reply index.bs:721]: Agreed this section does not make much sense right now.
It can be revived if there is WG consensus to move on with supporting audio.

File: index.bs [hunk 12]
```diff
@@ -630,7 +589,199 @@ self.onmessage = async function(e) {
...
+This interface had the generator for the MediaStreamTrack being an instance of
+a MediaStreamTrack rather than containing one.
+
+This interface can be shimmed on top of VideoTrackSource like this:
+
+<pre>
+
+Javascript goes here when written
+
+</pre>
```
[the remark this answers is not in the document]
Comment (alvestrand) [reply index.bs:708]: I don't agree, for entirely practical reasons: Leaving this shim in here will allow pepole to use the new API shape with existing browsers, letting them early-adopt the new API without waiting for its implementation.

[the remark this answers is not in the document]
Comment (alvestrand) [reply index.bs:708]: Hm. This can be left to an advice note when shipping the new interface in preference to the older one.
I can leave it out; the reason to include it would be to prove by code that the new interface can do everything the old interface could.

File: index.bs (hunk 3 above)
Comment (alvestrand) [reply index.bs:721]: This is about preserving previous considerations. This isn't code.

File: index.bs (hunk 6 above)
Comment (alvestrand) [reply index.bs:322]: Don't understand this comment. What are you suggesting?

**Comment by alvestrand** [comment]
I'm interested in getting this to the WG ASAP for the call for adoption, so would be inclined to minimize changes.
Question: In a couple of places, it's been suggested to "leave this in alvestrand/mediacapture-transform". This can't be done until the move/clone dance has completed, of course (this IS the alvestrand repo). But it makes sense after the move/clone dance has completed.

WDYT about augmenting the backwards compatibility section with an explicit pointer to "older version hosted at..."?

**Review by alvestrand — COMMENTED** [review]
Addressed a number of comments, and changed name from VideoTrackSource to VideoTrackGenerator (it is a special kind of source, not a generic source object).

Leaving the revision of examples to a later CL (Jan-Ivar will do that better than I can).

PTAL.

**Comment by youennf** [comment]
> WDYT about augmenting the backwards compatibility section with an explicit pointer to "older version hosted at..."?

It seems ok for the standard draft to acknowledge the initial non standard document it was based upon, for instance as a link in the introduction.
The non standard document can then document at will how to shim/feature detect/migrate from the non standard API to the standard API.

**Comment by jan-ivar** [comment]
Sorry for the late reply.

> WebIDL is a clearer exposition format than prose, and this is an informative section.

I think the WebIDL should at least be enclosed in a `<div class="example">` tag like [we do in mediacapture](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-constrainablepattern), otherwise it gets picked up by various W3C tools that generate all sorts of tests meant for standards track APIs.

But if we link to the non standard document as Youenn suggests, then I'd prefer we move this WebIDL there.

I'll do a PR for examples.

**Comment by jan-ivar** [comment]
> I'll do a PR for examples.

Added https://github.com/alvestrand/mediacapture-transform/pull/67

**Comment by foolip** [comment]
I came across this note in https://w3c.github.io/mediacapture-transform/#track-processor-interface:

> There is WG consensus that the interface should be exposed on DedicatedWorker. There is no WG consensus on whether or not the interface should not be exposed on Window.

Trying to figure out what that's about I found my way here. But is there an issue or something tracking this issue? That issue in short is that Chrome exposes `MediaStreamTrackProcessor` only on `Window` and the spec only on `DedicatedWorker`.

**Comment by dontcallmedom** [comment]
#23 is where this has been discussed
